src/FieldEncryptProcessEntities.php

```diff
@@ -0,0 +1,302 @@
+   *
+   * @param \Drupal\Core\Field\FieldItemListInterface $field
+   * @param string $op
+   * @param boolean $force if set, we don't check if encryption is enabled, we process the field anyway. This is used during batch processes.
+   */
+  protected function process_field(\Drupal\Core\Field\FieldItemListInterface $field, $op = 'encrypt', $force = FALSE) {
+    if (!is_callable([$field, 'getFieldDefinition'])){return;}
+
+    /**
+     * @var $definition \Drupal\Core\Field\BaseFieldDefinition
+     */
+    $definition = $field->getFieldDefinition();
+
+    if (!is_callable([$definition, 'get'])){
+      return;
+    }
+
+    /**
+     * Filter (blacklist) field types that don't work with encryption.
+     */
+    if (in_array($definition->get('field_type'), $this->blacklist_fields)) {
+      return;
+    }
+
+    /**
+     * @var $storage \Drupal\Core\Field\FieldConfigStorageBase
+     */
+    $storage = $definition->get('fieldStorage');
+    if (is_null($storage)) {
+      return;
+    }
+
+    /**
+     * If we are using the force flag, we always proceed.
+     * The force flag is used when we are updating stored fields.
+     */
+    if (!$force) {
+      /**
+       * Check if we are updating the field, in that case, skip it now (during
+       * the initial entity load.
+       */
+      if ($this->updatingStoredField === $definition->get('field_name')) {
+        return;
+      }
+
+      // Check if the field is encrypted.
+      $encrypted = $storage->getThirdPartySetting('field_encrypt', 'encrypt', FALSE);
+      if (!$encrypted) {
+        return;
+      }
+    }
+
+    /**
+     * @var $field \Drupal\Core\Field\FieldItemList
+     */
+    $field_value = $field->getValue();
+    foreach($field_value as &$value) {
+      // Process each of the sub fields that exits.
+      foreach($this->field_value_fields as $field_name) {
+        if(isset($value[$field_name])){
+          $value[$field_name] = $this->process_value($value[$field_name], $op);
+        }
+      }
+    }
+    // Set the new value.
+    // We don't need to update the entity because the field setValue does that already.
+    $field->setValue($field_value);
+  }
+
+  /**
+   * Process an entity to either encrypt or decrypt its fields.
+   *
+   * Both processes are very similar, so we bundle the field processing part.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   * @param string $op
+   */
+  protected function process_entity(\Drupal\Core\Entity\ContentEntityInterface $entity, $op = 'encrypt') {
+    // Make sure we can get fields.
+    if (!is_callable([$entity, 'getFields'])){
+      return;
+    }
+
+    foreach ($entity->getFields() as $field){
+      $this->process_field($field, $op);
+    }
+  }
+
+  /**
+   * Encrypt stored fields.
+   *
+   * This is performed when field storage settings are updated.
+   *
+   * @param $entity_type
+   * @param $field_name
+   */
+  public function encrypt_stored_field($entity_type, $field_name) {
```

Comment:
Is all of this documented in the readme? I'm having a hard time understanding how all these functions fit together